### DISCUSSION TITLE
Multiple Band Confusion

### DISCUSSION THREAD

**Discussion by rrowlands** [developmentseed/titiler#629] [Q&A] [answered]
I'm having issues getting Titiler to render properly when rendering a cog tiff with 6 bands. 

As you can see in this screenshot:
![image](https://user-images.githubusercontent.com/1326328/231539900-a1232679-9c8b-4621-b9ba-bddba3b4c438.png)

The colors are quite distorted. Here is what the ortho is supposed to look like:

![image](https://user-images.githubusercontent.com/1326328/231545294-6e9cccfe-6ab0-4369-af4a-e557dba47247.png)

I suspect that perhaps the additional bands are somehow modifying the rendering but I'm having trouble figuring out how exactly. My ortho has 6 bands as:

```
Band 1 Block=256x256 Type=UInt16, ColorInterp=Blue
  Description = Blue
  Min=0.000 Max=64585.000   Computed Min/Max=0.000,65535.000
  Minimum=0.000, Maximum=64585.000, Mean=17135.034, StdDev=18675.052
  Overviews: 6345x4111, 3172x2055, 1586x1027, 793x513, 396x256, 198x128
  Metadata:
    STATISTICS_APPROXIMATE=YES
    STATISTICS_MAXIMUM=64585
    STATISTICS_MEAN=17135.034051452
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=18675.052464637
    STATISTICS_VALID_PERCENT=100
Band 2 Block=256x256 Type=UInt16, ColorInterp=Green
  Description = Green
  Min=0.000 Max=63789.000   Computed Min/Max=0.000,65535.000
  Minimum=0.000, Maximum=63789.000, Mean=17786.620, StdDev=19242.059
  Overviews: 6345x4111, 3172x2055, 1586x1027, 793x513, 396x256, 198x128
  Metadata:
    STATISTICS_APPROXIMATE=YES
    STATISTICS_MAXIMUM=63789
    STATISTICS_MEAN=17786.619554924
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=19242.05863315
    STATISTICS_VALID_PERCENT=100
Band 3 Block=256x256 Type=UInt16, ColorInterp=Red
  Description = Red
  Min=0.000 Max=64208.000   Computed Min/Max=0.000,65535.000
  Minimum=0.000, Maximum=64208.000, Mean=17543.328, StdDev=19091.902
  Overviews: 6345x4111, 3172x2055, 1586x1027, 793x513, 396x256, 198x128
  Metadata:
    STATISTICS_APPROXIMATE=YES
    STATISTICS_MAXIMUM=64208
    STATISTICS_MEAN=17543.327769886
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=19091.902193138
    STATISTICS_VALID_PERCENT=100
Band 4 Block=256x256 Type=UInt16, ColorInterp=Gray
  Description = NIR
  Min=0.000 Max=65421.000   Computed Min/Max=0.000,65535.000
  Minimum=0.000, Maximum=65421.000, Mean=20173.965, StdDev=21353.978
  Overviews: 6345x4111, 3172x2055, 1586x1027, 793x513, 396x256, 198x128
  Metadata:
    STATISTICS_APPROXIMATE=YES
    STATISTICS_MAXIMUM=65421
    STATISTICS_MEAN=20173.964883207
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=21353.977824611
    STATISTICS_VALID_PERCENT=100
Band 5 Block=256x256 Type=UInt16, ColorInterp=Gray
  Description = Rededge
  Min=0.000 Max=65535.000   Computed Min/Max=0.000,65535.000
  Minimum=0.000, Maximum=65535.000, Mean=19570.810, StdDev=20874.293
  Overviews: 6345x4111, 3172x2055, 1586x1027, 793x513, 396x256, 198x128
  Metadata:
    STATISTICS_APPROXIMATE=YES
    STATISTICS_MAXIMUM=65535
    STATISTICS_MEAN=19570.810053662
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=20874.292767513
    STATISTICS_VALID_PERCENT=100
Band 6 Block=256x256 Type=UInt16, ColorInterp=Alpha
  Min=0.000 Max=255.000   Computed Min/Max=0.000,255.000
  Minimum=0.000, Maximum=255.000, Mean=124.160, StdDev=127.456
  Overviews: 6345x4111, 3172x2055, 1586x1027, 793x513, 396x256, 198x128
  Metadata:
    STATISTICS_APPROXIMATE=YES
    STATISTICS_MAXIMUM=255
    STATISTICS_MEAN=124.15956439394
    STATISTICS_MINIMUM=0
    STATISTICS_STDDEV=127.45623362536
    STATISTICS_VALID_PERCENT=100
```

And the same issue persists if I define bidx as 3,2,1. I am not exactly sure how Titiler determines which band is the alpha band, and I haven't figured out how to tell Titiler which band to use as alpha so it could be the case that it's using the wrong band?

I have noticed the following quirks regarding my ortho which, when "fixed", allow titiler to render it properly:
1. The bands are b,g,r,nir,g,a. Remove bands 4 and 5 and reorder as rgba
2. The alpha band is defined as data type UInt16, however it only ranges from 0-255, which would imply an actual data type of UInt8. Rescale from 0 to UInt16 max (65535).

Is there a way I can get Titiler to render this ortho correctly without pre-processing it as above using GDAL?

I have tested this on versions 0.11.0 and 0.5.1 of Titiler deployed on AWS Lambda.

**Comment by vincentsarago** [comment]
Thanks @rrowlands for the report 

This is definitely a bug but not in titiler. 

if we look at the `/preview` result:

https://titiler.xyz/cog/preview?url=https://osmre-uas-dev-public.s3.amazonaws.com/s/p/m/full/ortho/odm_orthophoto.cog.tif&bidx=3&bidx=2&bidx=1&rescale=0,65535&max_size=512

![preview-2](https://user-images.githubusercontent.com/10407788/231557594-637ff0ad-533c-4230-88a5-94f5450971f6.png)

It looks as expected. But when doing a `Tile` request we notice the weird `colours`. 

The difference between the `/preview` and the `/tiles` request is that rio-tiler will do a virtual re-projection using `rasterio.vrt.WarpedVRT` when doing the tile request while it will read the data simply when doing the preview request.

https://github.com/cogeotiff/rio-tiler/blob/b7ce316bdb69aac48b3792bf9d26ac634c9426e5/rio_tiler/reader.py#L129-L149

### Preview with and without reprojection
we can confirm this by applying a reprojection with the preview

<img width="815" alt="Screenshot 2023-04-12 at 9 01 25 PM" src="https://user-images.githubusercontent.com/10407788/231558619-0856b782-19cc-4b0e-b092-d898e9629d0e.png">

### WarpedVRT issue

<img width="814" alt="Screenshot 2023-04-12 at 9 13 02 PM" src="https://user-images.githubusercontent.com/10407788/231561020-afba334a-402e-4871-bcb7-b9d68743088d.png">

### Next Steps

Try to reproduce the issue with GDAL to determine if it's in rasterio or GDAL 🙏

**Comment by vincentsarago** [comment]
GDAL WarpedVRT seems fine, but we should verify using python GDAL 🤷 

<img width="818" alt="Screenshot 2023-04-12 at 9 21 42 PM" src="https://user-images.githubusercontent.com/10407788/231562867-88cf2187-1399-4225-b0cd-025aa40ae300.png">


I think the next steps would be to open an issue in rasterio with a minimal reproducible example

**Comment by justinlewis** [comment]
Thanks @vincentsarago! This is very helpful. We'll dig in further and see what we find.

**Reply by vincentsarago** [reply]
🤯 I think I found a possible fix. I didn't notice but there is an `src_alpha` option in Rasterio's WarpedVRT method (https://github.com/rasterio/rasterio/blob/main/rasterio/_warp.pyx#L833-L834) which is set to 0 by default...

Let me see if I can push a fix in rio-tiler

**Reply by vincentsarago** [reply, marked as answer]
alright I think I have a good idea of what's going on. 

There is a bug in rasterio in https://github.com/rasterio/rasterio/blob/d008e8aab38afd5d14d98499847754bd0eeaa501/rasterio/_warp.pyx#L978-L979  

I'll open a ticket there and try to submit a PR

**Reply by rrowlands** [reply]
Thanks for your help @vincentsarago !

**Reply by justinlewis** [reply]
@vincentsarago  Just to be clear, it sounds like you think that Rasterio bug is causing the issue and you can't update RioTiler until that bug in Rasterio is fixed? Just trying to make sure I understand where we're at so I can understand if we need to spend time helping to debug further.

**Reply by vincentsarago** [reply]
This is a rasterio bug and sadly there is no quick fix in rio-tiler for this 😭 

I've opened an issue https://github.com/rasterio/rasterio/issues/2808 and a PR https://github.com/rasterio/rasterio/pull/2809

Sadly we will have to wait to this to get merged and a new rasterio release to get published

**Reply by rrowlands** [reply]
I am removing the S3 file link since I can't make this data public unfortunately.

**Reply by vincentsarago** [reply]
oops sorry

**Reply by vincentsarago** [reply]
@rrowlands next time you could share `fake` version of the file using https://github.com/cogeotiff/rio-faux
